lutris/website
# Allow users to suggest game metadata changes

# TL;DR

- Fixes #98
- Add user functionality: *Submit change suggestions*
- Add admin functionality: *Review change suggestions*, *Accept change suggestions*, *Reject change suggestions* (features bulk-editing)
- I am grateful for feedback

# Screenshots

Here are some screenshots showing the new functionality (may be out of date).

<details>
<summary>Screenshots of the changes considering frontend use</summary>

![screenshot from 2017-10-25 17-31-04](https://user-images.githubusercontent.com/13337480/32007872-14810ef4-b9ab-11e7-9954-59c4cddd7aba.png)
**Figure 1**: New button in the top right corner to get to the change form.  

![screenshot from 2017-10-25 17-31-24](https://user-images.githubusercontent.com/13337480/32007873-14c6a180-b9ab-11e7-80a2-ed6a31974cfc.png)
**Figure 2**: The change form is an altered submission form. Most of the tooltips were not necessary anymore and the banner widget caused issues.  

![screenshot from 2017-10-25 17-31-54](https://user-images.githubusercontent.com/13337480/32007874-14eac6c8-b9ab-11e7-8f28-5bd6b8fc1007.png)
**Figure 3**: A new template to thank the user for the submission.  

</details>

<details>
<summary>Screenshots of the changes considering backend use</summary>

![screenshot from 2017-10-25 17-32-35](https://user-images.githubusercontent.com/13337480/32007875-150d8bf4-b9ab-11e7-9790-8000c36ebee0.png)
**Figure 4**: New filter in the admin panel to filter for *games only*, *changes for games only*, or *all*.  

![screenshot from 2017-10-25 17-32-59](https://user-images.githubusercontent.com/13337480/32007876-1534b896-b9ab-11e7-9abe-5be96497d01f.png)
**Figure 5**: Same view as Figure 4, but filtered for *changes for games only*. Note the column *actions* which links to a review page (see Figure 7).

![screenshot from 2017-10-28 15-15-53](https://user-images.githubusercontent.com/13337480/32134734-b381d110-bbf3-11e7-964d-5143105b8ab9.png)
**Figure 6**: Same view as Figure 4 and Figure 5, but with different data. Here, the filter is set to *all*, showing both games and change suggestions. Note that "Game Beta" has two change submissions pending as indicated in the *actions* column. Clicking on it, will direct to Figure 9.

<img alt="screenshot from 2017-10-25 17-33-21" src="https://user-images.githubusercontent.com/13337480/32007878-15b12b9c-b9ab-11e7-905e-16c9f4d11bf2.png" width="300"></img>
**Figure 7**: New custom admin template to review a submission. It lists the changes only and features buttons to accept or reject a submission.

![screenshot from 2017-10-28 15-18-04](https://user-images.githubusercontent.com/13337480/32134710-44fc05b2-bbf3-11e7-8781-595c2833a65e.png)
**Figure 8**: For staff members with the permission `game.change_game` a button will be displayed if there are change suggestions pending for this game. Clicking on it will redirect you to Figure 9.

![screenshot from 2017-10-28 13-57-50](https://user-images.githubusercontent.com/13337480/32134191-2e2fa6be-bbe8-11e7-81fa-e42e3df832e6.png)
**Figure 9**: An admin view to list all change suggestions for a given game. A similar view is available which lists all change suggestions for any game.

</details>

# Details

- Added a new column to the `games` table, called `change_for`, which is a foreign key to itself. If this field is not `NULL`, then the data row is a changeset for another game, namely the key it holds in `change_for` (*Reason: I wanted to reuse the existing table to not suffer from redundancy*).
- What a user can suggest for editing is controlled by methods in the `Game` model (*Reason: I personally prefer restrictive access control*).
- Change submission rows will be deleted when they get rejected or accepted. If they get accepted, the changes will be applied to the game it references through `change_for`.
- The slugs column had to be changed to allow for a `NULL` value. When a change is created, the slug will be set to `NULL`, since it is not relevant (*Reason: Slugs are `UNIQUE` it wouldn't have worked otherwise, except for random slugs, which would be hacky*).
- Even if the name of a game is changed, the slug will remain the same (*Reason: URL's would change otherwise, leading to 404's if users bookmarked something, SEO issues, ... Slugs should, however, be changed if there's a typo in them or something similar. This can still be done via the conventional admin interface*).

# Misc

- [Removed a part of a Unit test for Game model which had no obvious testing purpose and lead to errors in conjunction with other tests](https://github.com/lutris/website/pull/101/files#diff-c47850c6f1f59fc1ecad6255eeab9b5eL22). You can tell me what this test was supposed to do and we can add a better test which will not error out.
- Hid the button `Edit` on the game details view if the staff member does not have the permission `games.change_game`

# Documentation

- Added documentations for all new modules, functions and methods
- Inline documentations

# Testing

- Form to submit change suggestions (`forms.GameEditForm`)
  - Ensure that a user cannot submit an unchanged form
  - Ensure that a user can submit valid changes
  - Ensure that a user cannot submit invalid changes
- Game model (`models.Game`)
  - Ensure that the form model properly represents the relevant attributes (`get_change_model`)
  - Ensure that `get_changes` properly lists all differences
  - Ensures that `apply_changes` persists the changes


# ToDo's

- [x] Ensure that no other views are effected by the new column, i.e. that other functions filter out games whose `change_for` field is not `NULL`
- [x] Honor permissions (`change_game`)

# Future work / Backlog

(I don't want to bloat this MR, so I decided to not implement these for now. Let me know if you find one of those mandatory)

- QoL - Review game submission: Make fields editable, add reset (to apply partial changes, fix typos, ....) 
  - **Reason**: Has alternative
  - Change suggestions can be edited in the admin panel anyway. Cumbersome, but it works
- EMail notifications for both staff and user (as feedback)
  - **Reason**: Requirements unclear
  - Do we want this? (Will get pretty verbose for the staff)
  - For games which were curated very carefully, we could implement a `read-only` or `locked` column, which prevents users from suggesting further changes. This can reduce the workload for admins.
- Banner cannot be changed at the moment due to a bug in `ImageField`. Have not found a solution for it yet...
  - **Reason**: Estimated effort high / Requirements unclear
